src/main/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGenerator.java

```diff
@@ -0,0 +1,217 @@
+        DetailAST curNode = root;
+        while (curNode.findFirstToken(TokenTypes.IDENT) == null) {
+            DetailAST toVisit = curNode.getFirstChild();
+            while (toVisit == null) {
+                toVisit = curNode.getNextSibling();
+                if (toVisit == null) {
+                    curNode = curNode.getParent();
+                }
+            }
+
+            if (curNode == root.getParent()) {
+                curNode = null;
+                break;
+            }
+
+            curNode = toVisit;
+        }
+        return curNode;
+    }
+
+    /**
+     * Finds if {@code DetailAst} element is unique, across all elements or across siblings.
+     * @param ast {@code DetailAST} ast element
+     * @return if {@code DetailAst} element is unique
+     */
+    private boolean isUniqueAst(DetailAST ast) {
+        boolean isOnly = true;
+        final DetailAST identAst = ast.findFirstToken(TokenTypes.IDENT);
+
+        if (identAst == null) {
+            DetailAST cur;
+            if (ast.getParent() == null) {
+                cur = ast;
+                while (cur.getPreviousSibling() != null) {
+                    cur = cur.getPreviousSibling();
+                }
+            }
+            else {
+                cur = ast.getParent().getFirstChild();
+            }
+
+            for (; cur != null;
+                 cur = cur.getNextSibling()) {
+                if (cur != ast && cur.getType() == ast.getType()) {
+                    isOnly = false;
+                    break;
+                }
+            }
+        }
+        return isOnly;
+    }
+
+    /**
+     * Returns list of nodes matching defined line and column number.
+     * @param root {@code DetailAST} root ast
+     * @return list of nodes matching defined line and column number
+     */
+    private List<DetailAST> getMatchingAstElements(DetailAST root) {
+        final List<DetailAST> result = new ArrayList<>();
+        DetailAST curNode = root;
+        while (curNode != null && curNode.getLineNo() <= lineNumber) {
+            if (curNode.getLineNo() == lineNumber && curNode.getColumnNo() == columnNumber
+                    && curNode.getType() != TokenTypes.IDENT) {
+                result.add(curNode);
+            }
+            DetailAST toVisit = curNode.getFirstChild();
+            while (curNode != null && toVisit == null) {
+                toVisit = curNode.getNextSibling();
+                if (toVisit == null) {
+                    curNode = curNode.getParent();
+                }
+            }
+
+            curNode = toVisit;
+        }
+        return result;
+    }
+
+    /**
+     * Returns full xpath query for given ast element.
+     * @param root {@code DetailAST} root element
+     * @param ast {@code DetailAST} ast element
+     * @return full xpath query for given ast element
+     */
+    private static String getXpathQuery(DetailAST root, DetailAST ast) {
+        String result = "";
```

Comment:
IDEA says that the string should be replaced with StringBuilder as you concatenate string in the loop. Use insert method of StringBuilder to insert the string in the 0 position